### PR TITLE
[SPARK-32567][SQL][FOLLOWUP] Fix CompileException when code generate for FULL OUTER shuffled hash join

### PR DESCRIPTION
**Pull request by roryqi**
### What changes were proposed in this pull request?
Add `throws IOException` to `consumeFullOuterJoinRow` in ShuffledHashJoinExec

### Why are the changes needed?
pr #34444 add code-gen for full outer shuffled hash join. If we don't have this patch,   when the dataframes are `fullter outer` shuffled hash joined, and then aggregate the results.  the dataframes will throw a `CompileException`. 
For example:
```scala
    val df1 = spark.range(5).select($"id".as("k1"))
    val df2 = spark.range(10).select($"id".as("k2"))
    df1.join(df2.hint("SHUFFLE_HASH"), $"k1" === $"k2", "full_outer").count()
```
The dataframe will throw an Exception which is as follows:
```
23:28:19.079 ERROR org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 175, Column 1: Thrown exception of type "java.io.IOException" is neither caught by a "try...catch" block nor declared in the "throws" clause of the declaring function
org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 175, Column 1: Thrown exception of type "java.io.IOException" is neither caught by a "try...catch" block nor declared in the "throws" clause of the declaring function
	at org.codehaus.janino.UnitCompiler.compileError(UnitCompiler.java:12021)
	at org.codehaus.janino.UnitCompiler.checkThrownException(UnitCompiler.java:9801)
	at org.codehaus.janino.UnitCompiler.checkThrownExceptions(UnitCompiler.java:9720)
	at org.codehaus.janino.UnitCompiler.findIMethod(UnitCompiler.java:9163)
	at org.codehaus.janino.UnitCompiler.compileGet2(UnitCompiler.java:5055)
```
### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add a new UT

### CODE REVIEW

**Comment by viirya** [comment]
cc @c21

**Comment by viirya** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala [hunk 1]
```diff
@@ -212,6 +212,13 @@ class WholeStageCodegenSuite extends QueryTest with SharedSparkSession
       Seq(Row(0, 0, 0), Row(1, 1, null), Row(2, 2, 2), Row(3, 3, null), Row(4, 4, null),
         Row(null, 5, null), Row(null, 6, null), Row(null, 7, null), Row(null, 8, null),
         Row(null, 9, null), Row(null, null, 1)))
+
+    // test join with Aggregate
+    val joinWithAggregateDF = df1.join(df2.hint("SHUFFLE_HASH"), $"k1" === $"k2", "full_outer")
+    assert(joinWithAggregateDF.queryExecution.executedPlan.collect {
+      case WholeStageCodegenExec(_ : ShuffledHashJoinExec) => true
+    }.size === 1)
+    assert(joinWithAggregateDF.count() === 10)
```
Comment (c21) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala:221]: I think we can just do

```
assert(joinUniqueDF.count() === 10)
```

**Review by c21 — APPROVED** [review]
Thanks @jerqi for fixing the issue! I was about to fix it as I realized the bug when working on FULL OUTEr sort merge join code-gen.

For other reviewers: the `throws java.io.IOException` needs to be added here, because some operators in downstream can define method `throws java.io.IOException` when consuming the input rows. Example in hash aggregate - https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala#L221 . So here the bug can happen when there's an aggregate after shuffled hash join code-gen-ed together in same stage.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Thanks for fixing this @jerqi and @c21 for reviewing it.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala (hunk 1 above)
Comment (jerqi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala:221]: Ok, I will fix. Thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala [hunk 2]
```diff
@@ -212,6 +213,7 @@ class WholeStageCodegenSuite extends QueryTest with SharedSparkSession
       Seq(Row(0, 0, 0), Row(1, 1, null), Row(2, 2, 2), Row(3, 3, null), Row(4, 4, null),
         Row(null, 5, null), Row(null, 6, null), Row(null, 7, null), Row(null, 8, null),
         Row(null, 9, null), Row(null, null, 1)))
+
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala:216]: nit: unnecessary change.

**Comment by viirya** [comment]
Last commits are test-only and GA was passed. Merging to master.

**Comment by jerqi** [comment]
thanks @viirya @c21
